Adding a convenient Docker

bruAristimunha opened the pull request:
This pull request adds a Dockerfile to the MOABB repository, allowing users to easily build and run the MOABB code in a containerized environment. 

The Dockerfile is based on one of the latest Pytorch images from Nvidia and installs all necessary dependencies for running MOABB. This pull request also creates a convenient script to include instructions for building and running MOABB using Docker.

I understand that the Nvidia image is the best choice for performance, and for the plan to add deep learning models in moabb, but this choice may add the step of logging into the Nvidia/torch hub with Docker.

[WIP] Script to run the moabb/run.py
[WIP] Increase description

bruAristimunha commented:
Hey there! I've made some changes to the script and Dockerfile that should make them a lot more readable and easy to work with. In the script, I've defined all the variables in one place and used string concatenation to make the command more compact. I've also broken the command into multiple lines for better readability.

For the Dockerfile, I've made some changes to reduce the final size, by chaining multiple commands in one line. I've also added an environment variable and copied the current directory and a requirements file to the container, and set the working directory.

All these changes should make the script and Dockerfile a lot more accessible for everyone to understand and work with. Let me know if you have any questions!

I will run the Dockerfile in my infra to see if I got the same results as report in wiki.

bruAristimunha commented:
Hi @Div12345,

Can you build/run/reproduce the docker script and see if everything is working?

sylvchev commented on docker/meta_requirements.txt:
```diff
@@ -0,0 +1,2 @@
+mne
+moabb
```
I think you could use a git link, to avoid waiting for a PyPi release: https://stackoverflow.com/a/35998253

bruAristimunha replied:
Excellent suggestion =) I loved!

sylvchev commented on docker/run_docker.sh:
```diff
@@ -0,0 +1,28 @@
+set -ex # Enable 'set -e' (exit on error) and 'set -x' (debugging) options
+
+TAG=moabb
+MOUNT_POINT="/home/bru/PycharmProjects/moabb/dataset/:/root/mne_data/"
```
Set as argument?

bruAristimunha replied:
Done =)

sylvchev commented on docker/run_docker.sh:
```diff
@@ -0,0 +1,28 @@
+
+TAG=moabb
+MOUNT_POINT="/home/bru/PycharmProjects/moabb/dataset/:/root/mne_data/"
+PIPELINE="/workdir/pipelines/"
```
We should copy pipelines directory from git or a mirror site.

bruAristimunha replied:
This is an interesting question. Every time we build Docker, we copy a version of moabb into the image that theoretically has the most up-to-date pipeline.

However, every time we run Docker, we don't necessarily pull the latest pipeline version, which would involve cloning the repository on every run.

Some possible solutions:

1. Integrate the build and run scripts to ensure we are always synchronized with the latest version; 

Note: we may have to change the build script to be self-contained (no hard preferences).

2. Add a small bash script to download the moabb URL before running the run.

3. Change the run.py script to accept a link instead of a path and force that to make it download a folder.

What do you prefer @sylvchev?

bruAristimunha replied:
number 2.

bruAristimunha commented on moabb/run.py:
```diff
@@ -97,6 +107,9 @@ def parser_init():
     options = parser.parse_args()
 
+    if options.mne_data is not None:
+        mne.set_config("MNE_DATA", options.mne_data)
```
set_download_dir

bruAristimunha commented:
Hi @Div12345 and @sylvchev!

Minor update, I implemented all the points.

I managed to run on the server to that I have access. In the test with one dataset (Cho), something is running. I left it running to download all the datasets, and then I will do a small stress test to run with as much CPU as possible.

Can you check on your side, @Div12345 or @sylvchev? I added the instruction sequence in the README.

sylvchev reviewed (approved):
Everything is working fine on the cluster and on my machine.

sylvchev commented:
Thank you so much @bruAristimunha for your work!
